File: packages/typescript-core/src/signer.ts
import { signingPayload } from "./crypto.js";
import { toHex } from "./util.js";

/**
 * How a `/partial-decrypt` request is authorized. `substrate`: an sr25519
 * `MultiSignature` (every signer shipped here). `ethereum`: EIP-712, accepted by
 * the wire protocol but only via a custom {@link Signer}.
 */
export type AuthScheme = "substrate" | "ethereum";

/** Auth fields attached to a `/partial-decrypt` request. */
export interface RequestAuth {
  auth: AuthScheme;
  /** `0x` + SCALE `AccountId` (Substrate) or empty (Ethereum). */
  requester: string;
  /** `0x` + SCALE `MultiSignature` (Substrate) or empty (Ethereum). */
  signature: string;
  eth_address?: string;
  valid_until?: number;
  eth_signature?: string;
}

/** Per-request context handed to a {@link Signer}. One request targets one node. */
export interface SigningRequest {
  secretId: bigint;
  subset: number[];
  /** The 1-based `dkg_index` of the node this request is addressed to. */
  recipientIndex: number;
  blockHash: Uint8Array;
  validUntil?: number;
}

/**
 * Authorizes a request without exposing its key to the SDK; the implementor
 * owns the framing (needed for EIP-712, which has no substrate account and
 * signs typed data). Mirrors Rust `matter_sdk::Signer`.
 */
export interface Signer {
  authScheme(): AuthScheme;
  authorize(req: SigningRequest): Promise<RequestAuth>;
}

/**
 * An on-chain identity plus raw-byte signing: all an HSM, KMS, or remote signer
 * must provide. The SDK derives all framing. {@link ApiKey} implements this.
 */
export interface KeySigner {
  readonly accountId: Uint8Array;
  sign(message: Uint8Array): Uint8Array | Promise<Uint8Array>;
}

/** SCALE enum index of `MultiSignature::Sr25519` (`Ed25519=0, Sr25519=1, Ecdsa=2`). */
const MULTISIGNATURE_SR25519 = 0x01;

/** Length of a raw sr25519 signature, before `MultiSignature` framing. */
const SR25519_SIGNATURE_BYTES = 64;

const ACCOUNT_ID_BYTES = 32;

/**
 * Frame a signature over `req` as Substrate auth fields: an sr25519 signature over
 * the canonical signing payload, wrapped as a SCALE `MultiSignature`, with the raw
 * `AccountId32` as `requester`. Every key-backed signer routes through here.
 */
export async function partialDecryptAuth(
  signer: KeySigner,
  req: SigningRequest,
): Promise<RequestAuth> {
  const { accountId } = signer;
  if (accountId.length !== ACCOUNT_ID_BYTES) {
    throw new Error(`accountId must be ${ACCOUNT_ID_BYTES} bytes, got ${accountId.length}`);
  }
  const payload = signingPayload(req.secretId, req.subset, req.blockHash, req.recipientIndex);
  const sig = await signer.sign(payload);
  if (sig.length !== SR25519_SIGNATURE_BYTES) {
    throw new Error(`sr25519 signature must be ${SR25519_SIGNATURE_BYTES} bytes, got ${sig.length}`);
  }
  const multisig = new Uint8Array(1 + SR25519_SIGNATURE_BYTES);
  multisig[0] = MULTISIGNATURE_SR25519;
  multisig.set(sig, 1);
  return { auth: "substrate", requester: toHex(accountId), signature: toHex(multisig) };
}

/** Adapt a {@link KeySigner} (API key, HSM/KMS client, wallet) into a {@link Signer}. */
export function keySigner(signer: KeySigner): Signer {
  return {
    authScheme: () => "substrate",
    authorize: (req) => partialDecryptAuth(signer, req),
  };
}

/**
 * Build a Substrate signer from a 32-byte account id and a `sign` callback that
 * returns the 64-byte sr25519 signature over the payload. The key never enters
 * the SDK. Equivalent to `keySigner({ accountId, sign })`.
 */
export function substrateSigner(
  accountId: Uint8Array,
  sign: (payload: Uint8Array) => Uint8Array | Promise<Uint8Array>,
): Signer {
  if (accountId.length !== ACCOUNT_ID_BYTES) throw new Error("accountId must be 32 bytes");
  return keySigner({ accountId, sign });
}

File: packages/go/mattersdk/signer.go
package mattersdk

import "fmt"

// SCALE enum index of MultiSignature::Sr25519 (Ed25519=0, Sr25519=1, Ecdsa=2).
const multisignatureSr25519 = 0x01

// Length of a raw sr25519 signature, before MultiSignature framing.
const sr25519SignatureBytes = 64

// RequestAuth is the auth fields attached to a /partial-decrypt request. The
// Ethereum fields are set only by an Ethereum-auth (EIP-712) Signer; the
// Substrate path leaves them nil.
type RequestAuth struct {
	Auth      string `json:"auth"`
	Requester string `json:"requester"`
	Signature string `json:"signature"`
	// EthAddress is the 0x-prefixed Ethereum address that signed.
	EthAddress *string `json:"eth_address,omitempty"`
	// ValidUntil is the unix time after which the signature is refused.
	ValidUntil *uint64 `json:"valid_until,omitempty"`
	// EthSignature is the 0x-prefixed EIP-712 signature.
	EthSignature *string `json:"eth_signature,omitempty"`
}

// Signer authorizes a /partial-decrypt request without exposing its key.
// recipientIndex is the target node's 1-based dkg_index; it is signed so a
// signature cannot be replayed to a peer. Called once per node in the quorum.
type Signer interface {
	AuthScheme() string
	Authorize(secretID SecretID, subset []uint64, recipientIndex uint64, blockHash [32]byte) (RequestAuth, error)
}

type substrateSigner struct {
	requester string
	sign      func([]byte) ([]byte, error)
}

// SubstrateSigner builds a Signer from a 32-byte account id and an sr25519 sign
// callback, which receives the canonical payload and returns the 64-byte
// signature. The key never enters the SDK; use ApiKey for an in-process key.
func SubstrateSigner(accountID []byte, sign func([]byte) ([]byte, error)) (Signer, error) {
	if len(accountID) != 32 {
		return nil, fmt.Errorf("accountID must be 32 bytes, got %d", len(accountID))
	}
	return &substrateSigner{requester: toHex(accountID), sign: sign}, nil
}

func (s *substrateSigner) AuthScheme() string { return "substrate" }

func (s *substrateSigner) Authorize(secretID SecretID, subset []uint64, recipientIndex uint64, blockHash [32]byte) (RequestAuth, error) {
	payload, err := SigningPayload(secretID.Bytes(), subset, blockHash, recipientIndex)
	if err != nil {
		return RequestAuth{}, err
	}
	sig, err := s.sign(payload)
	if err != nil {
		return RequestAuth{}, err
	}
	if len(sig) != sr25519SignatureBytes {
		return RequestAuth{}, fmt.Errorf("sr25519 signature must be %d bytes, got %d", sr25519SignatureBytes, len(sig))
	}
	multisig := append([]byte{multisignatureSr25519}, sig...)
	return RequestAuth{Auth: "substrate", Requester: s.requester, Signature: toHex(multisig)}, nil
}
